<a href="https://colab.research.google.com/github/NatalieGo/MFTI_hackathon_1/blob/main/%D0%A0%D0%B0%D0%B7%D0%B1%D0%BE%D1%80_json_isp.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import json
from IPython.display import display, HTML
data = pd.read_json('data/RU_dataset.json')
display(data)
display(data.info())

0            1
0      Бывшая жена угрожает самоубийством Недавно я н...      suicide
1      Я странный? На меня не влияют комплименты, есл...  non-suicide
2      Наконец-то 2020 год почти закончился... Так чт...  non-suicide
3       мне нужна помощь, просто помоги мне, я так плачу      suicide
4      Я так растеряна. Здравствуйте, меня зовут Адам...      suicide
...                                                  ...          ...
47992  Я только что увидел забавный мем, прежде чем п...      suicide
47993  Я боюсь, мне некому рассказать. У меня инфекци...  non-suicide
47994  Педофилия, членовредительство и мысли о самоуб...      suicide
47995  Я больше не могу справляться. Почти 2 месяца н...      suicide
47996  Мне нужен кто-то, кто сделает сюрприз моему дв...  non-suicide

[47997 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47997 entries, 0 to 47996
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       47997 non-null  object
 1   1       47997 non-null  object
dtypes: object(2)
memory usage: 750.1+ KB


None

В данном датасете информация о 47997 сообщениях с Raddit, которые размечены по 2м категориям: «non-suicide» и «suicide»
Содержит 47997 строк, 2 столбца («0» и «1» ), оба столбца имеют тип object. для того, что бы оценить наличие лишних символов и корректность разметки, преобразуем файл в CSV формат. Так же переименуем столбцы.

In [ ]:
data = data.rename(columns={0: 'Сообщение', 1: 'Самоубийство'})
display(data)


Сообщение Самоубийство
0      Бывшая жена угрожает самоубийством Недавно я н...      suicide
1      Я странный? На меня не влияют комплименты, есл...  non-suicide
2      Наконец-то 2020 год почти закончился... Так чт...  non-suicide
3       мне нужна помощь, просто помоги мне, я так плачу      suicide
4      Я так растеряна. Здравствуйте, меня зовут Адам...      suicide
...                                                  ...          ...
47992  Я только что увидел забавный мем, прежде чем п...      suicide
47993  Я боюсь, мне некому рассказать. У меня инфекци...  non-suicide
47994  Педофилия, членовредительство и мысли о самоуб...      suicide
47995  Я больше не могу справляться. Почти 2 месяца н...      suicide
47996  Мне нужен кто-то, кто сделает сюрприз моему дв...  non-suicide

[47997 rows x 2 columns]

In [ ]:
#Идентификация пропущенных значений (NaN, NULL, пустых строк и т. д.) в датафрейме
filtered_colums = data.eq("").sum()
display(filtered_colums)
print(data.info())
data.isnull().sum()

Сообщение       0
Самоубийство    0
dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47997 entries, 0 to 47996
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Сообщение     47997 non-null  object
 1   Самоубийство  47997 non-null  object
dtypes: object(2)
memory usage: 750.1+ KB
None


Сообщение       0
Самоубийство    0
dtype: int64

In [ ]:
#Посчитаем сколько постов относятся к  «Пост о потенциальном самоубийстве»
data['Самоубийство'].value_counts()

Самоубийство
non-suicide    24081
suicide        23916
Name: count, dtype: int64

Посчитаем сколько постов сигнализируют о риске суицида, получаем что : «non-suicide» - 24081  и «suicide» - 23916. Видим равномерное распределение. Это плюс для нашей модели. Удаляем лишние символы. Удаляем неинформативные строки, состоящие только из цифр и длинной менее 5 символов. На этом этапе выводим 50 первых строк для оценки корректности разметки.

In [ ]:
# используем IPython.display для красивого вывода всего текста в таблице
from IPython.display import display, HTML
data['Сообщение'] = data['Сообщение'].replace(r'http[s]?://\S+|@\w+|/|&\w+|ð\w+|¶+|RT|aâ¦|@|w\w+|:|#\w+|_|-|ð|ó|¦|§|®|¥|¢|´|¿|ï½', '', regex=True)
data['Сообщение'] = data['Сообщение'].str.replace(r'^\,+', '', regex=True)
data['Сообщение'] = data['Сообщение'].str.replace(r'\n', '', regex=True)
data['Сообщение'] = data['Сообщение'].str.replace(r'XD', '', regex=True)
data['Сообщение'] = data['Сообщение'].str.replace(r'<3', '', regex=True)
data['Сообщение'] = data['Сообщение'].str.lower()
data['Сообщение'] = data['Сообщение'].str.lstrip()
data = data[data['Сообщение'].apply(lambda x: len(x) > 5)] #удаляем неинформативные строки длиной мененее пяти символов
data = data[data['Сообщение'].apply(lambda x: x.isdigit() == False)] #удаляем строки состоящие только из цифр
display(HTML(data.tail(10).to_html()))

In [ ]:
#Снова посчитаем данные о количестве постов
data['Самоубийство'].value_counts()

Самоубийство
non-suicide    24080
suicide        23915
Name: count, dtype: int64

Осталось 24077 сообщений не сигнализирующих о самоубийстве и 23914 сообщений сигнализирующих о самоубийстве. Сохраним новый вариант дата сета.

In [ ]:
data.to_csv('data/ru_Sui_new_isp_5.csv', index=False, sep='\t')